packages/api/src/car.js

```diff
@@ -19,44 +22,86 @@ export async function carHead(request, env, ctx) {
-    if (!res.ok) {
-      // bail early. dont cache errors.
-      return res
-    }
-    // TODO: these headers should be upstreamed to ipfs impls
-    // without the content-disposition, firefox describes them as DMS files.
-    const headers = {
-      'Cache-Control': 'public, max-age=31536000',  // max max-age is 1 year
-      'Content-Type': 'application/car',
-      'Content-Disposition': `attachment; filename="${cid}.car"`,
-    }
-    // // compress if asked for? is it worth it?
-    // if (request.headers.get('Accept-Encoding').match('gzip')) {
-    //   headers['Content-Encoding'] = 'gzip'
-    // }
-    res = new Response(res.body, { ...res, headers })
-    ctx.waitUntil(cache.put(request, res.clone()))
+
+  if (res) {
+    return res
   }
+
+  const {
+    params: { cid },
+  } = request
+  // gateway does not support `carversion` yet.
+  // using it now means we can skip the cache if it is supported in the future
+  const url = new URL(`/api/v0/dag/export?arg=${cid}&carversion=1`, GATEWAY)
+  console.log(url.toString())
+  res = await fetch(url, { method: 'POST' })
+  if (!res.ok) {
+    // bail early. dont cache errors.
+    return res
+  }
+  // TODO: these headers should be upstreamed to ipfs impls
+  // without the content-disposition, firefox describes them as DMS files.
+  const headers = {
+    'Cache-Control': 'public, max-age=31536000',  // max max-age is 1 year
+    'Content-Type': 'application/car',
+    'Content-Disposition': `attachment; filename="${cid}.car"`,
+  }
+  // // compress if asked for? is it worth it?
+  // if (request.headers.get('Accept-Encoding').match('gzip')) {
+  //   headers['Content-Encoding'] = 'gzip'
+  // }
+  res = new Response(res.body, { ...res, headers })
+  ctx.waitUntil(cache.put(request, res.clone()))
+
   return res
 }
 
 /**
+ * Post a CAR file.
+ *
  * @param {import('./user').AuthenticatedRequest} request
+ * @param {import('./env').Env} env
  */
-export async function carPost(request, env, ctx) {
-  const { user, authToken } = request.auth
-  console.log('User ID', user._id)
-  console.log('Auth Token ID', authToken && authToken._id)
-  return new JSONResponse({})
+export async function carPost(request, env) {
+  const { _id } = request.auth.authToken
+  const { headers } = request
+
+  let name = headers.get('x-name')
+  if (!name || typeof name !== 'string') {
+    name = `Upload at ${new Date().toISOString()}`
+  }
+
+  const blob = await request.blob()
+
+  // Ensure car blob.type is set; it is used by the cluster client to set the foramt=car flag on the /add call.
+  const content = blob.slice(0, blob.size, 'application/car')
+
+  const { cid } = await env.cluster.add(content, {
+    metadata: { size: content.size.toString() },
+    // When >2.5MB, use local add, because waiting for blocks to be sent to
+    // other cluster nodes can take a long time. Replication to other nodes
+    // will be done async by bitswap instead.
+    local: blob.size > LOCAL_ADD_THRESHOLD,
```

Comment:
Why not `--local` add everything? with a 3 of 3 replication, we could get stuck waiting for content to add to all 3 nodes if 1 node is backed up

Reply:
I've asked hector what happens...

Reply:
> During the add cluster does block-puts for every block. As long as block-put works in 1 destination it keeps going and the adding does not error.
> If block-put fails in one place once, then that place is not tried again for subsequent blocks.
> At the end of the adding process you can be sure at least 1 place has all the blocks.

...so it sounds like we don't get stuck :)